### PR TITLE
kvutils: Authorization and config submissions

### PR DESCRIPTION
**Pull request by dajmaki**
Implements authorization of submissions and expands configuration submissions.
Expands kvutils testing.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 1]
```diff
@@ -279,6 +314,7 @@ message DamlStateKey {
     DamlCommandDedupKey command_dedup = 3;
     string party = 4;
     DamlContractKey contract_key = 5;
+    google.protobuf.Empty current_configuration = 6;
```
Comment (mziolekda) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: configuration_key?

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteConfigService.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package com.daml.ledger.participant.state.v1
+
+import java.util.concurrent.CompletionStage
+
+import com.digitalasset.daml.lf.data.Time.Timestamp
+
+trait WriteConfigService {
```
Comment (mziolekda) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteConfigService.scala]: Is this interface mandatory for all X's? I am thinking especially here about Daml-on-SQL and Canton.

**Comment by dajmaki** [comment]
this actually had bunch of commits that already were in master.. there wasn't actually _that_ much churn, but enough that I'll split this up.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteConfigService.scala (hunk 2 above)
Comment (dajmaki) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteConfigService.scala]: they can choose to return "NotImplemented" error.

Comment (mziolekda) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteConfigService.scala]: We should write that in the comment then

Comment (dajmaki) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteConfigService.scala]: commented.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 3]
```diff
@@ -364,3 +426,8 @@ message DamlContractKeyState {
   // If unset the key is inactive.
   DamlContractId contract_id = 1;
 }
+
+// Participant not authorized to submit the request, a common rejection reason.
```
Comment (mziolekda) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: I am not sure what pulling up just one error buys us. It seems we lose on the type safety side by gaining some ability to write reusable error handlers.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 4]
```diff
@@ -189,32 +201,85 @@ message DamlPackageUploadRejectionEntry {
...
-
     InvalidPackage invalid_package = 3;
     ParticipantNotAuthorized participant_not_authorized = 4;
   }
 }
 
-message DamlConfigurationEntry {
-  DamlTimeModel time_model = 1;
+// Configuration change request to change the ledger configuration.
+message DamlConfigurationSubmission {
```
Comment (mziolekda) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: It seems configuration needs a submission id, just like party allocation and package upload, else it is difficult to match responses with requests, generation number is not globally unique and 2 participants will decide on the same subsequent number, opening up the possibility of mixing up the responses.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 3 above)
Comment (dajmaki) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: It also isn't going to play nicely if we modularize this. I'll revert the changes.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 4 above)
Comment (dajmaki) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: I considered it, but felt that it doesn't need to be that complex due to the configuration changes being very rare and almost invariably coming from a single source (by design configuration can also only be changed by one "admin" participant). Though on the other hand if we add things to the configuration that may be tweaked more regularly or from multiple sources at once we might want to consider just adding it in. 

Do you think we still should add it?

Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: I would add it for consistency sake

Comment (dajmaki) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: cool, I'll add it.

**Pull request merged**